src/bin/bat/app.rs

```diff
@@ -125,36 +125,43 @@ impl App {
             }
         }
 
-        let maybe_term_width = self.matches.value_of("terminal-width").and_then(|w| {
-            if w.starts_with('+') || w.starts_with('-') {
-                // Treat argument as a delta to the current terminal width
-                w.parse().ok().map(|delta: i16| {
-                    let old_width: u16 = Term::stdout().size().1;
-                    let new_width: i32 = i32::from(old_width) + i32::from(delta);
-
-                    if new_width <= 0 {
-                        old_width as usize
-                    } else {
-                        new_width as usize
-                    }
-                })
-            } else {
-                w.parse().ok()
-            }
-        });
+        let maybe_term_width = self
+            .matches
+            .get_one::<String>("terminal-width")
+            .and_then(|w| {
+                if w.starts_with('+') || w.starts_with('-') {
+                    // Treat argument as a delta to the current terminal width
+                    w.parse().ok().map(|delta: i16| {
+                        let old_width: u16 = Term::stdout().size().1;
+                        let new_width: i32 = i32::from(old_width) + i32::from(delta);
+
+                        if new_width <= 0 {
+                            old_width as usize
+                        } else {
+                            new_width as usize
+                        }
+                    })
+                } else {
+                    w.parse().ok()
+                }
+            });
 
         Ok(Config {
             true_color: is_truecolor_terminal(),
-            language: self.matches.value_of("language").or_else(|| {
-                if self.matches.is_present("show-all") {
-                    Some("show-nonprintable")
-                } else {
-                    None
-                }
-            }),
-            show_nonprintable: self.matches.is_present("show-all"),
+            language: self
+                .matches
+                .get_one::<String>("language")
+                .map(|s| s.as_str())
```

Comment:
Just curious: this seems quite a bit more verbose than what we had before. Also, it looks like it now requires some dynamically allocated String, whereas before we directly got access to a `&str` or `&OsStr`. What were the reasons for this API change?

Reply:
> What were the reasons for this API change?

Clap now allows you to store the parsed value (PathBuf, usize, etc) directly in the `ArgMatches`.  `get_one::<T>()` accepts what the type is stored inside of `ArgMatches`.

I was a bit concerned with the extra ceremony around this but we decided it was worth it
- For the extra features this enabled
  - Avoids double parsing
  - Bypasses problems we had with the `value_of_t` API we previously maintained
  - We're able to integrate flags directly into clap without a lot of back and forth (previously clap treated them special by not storing any value but separately tracking the presence and the number of occurrences, now its directly supported allowing integration with other clap features)
- It was limited to the Builder API where ease of use is a lower priority compared to the Derive API

> If we need to pass a type argument to .get_one, is it possible to directly .get_one::<&str>?

I considered playing with trait to allow mapping `&str` to `String` but I was concerned with usability issues
- If the user specified the type (`arg.value_parser(value_parser!(usize)))`, then that type wouldn't match with what is used at the access
- When we report an assert on type mismatches, we would be suggesting to the user the owned type and not the borrowed type

> If not, can .map(|s| s.as_str()) be replaced by .as_deref()?

`&String` does not deref to `&str` but to itself

```
error[E0308]: mismatched types
   --> src/bin/bat/main.rs:231:43
    |
231 |         bat::config::get_pager_executable(app.matches.get_one::<String>("pager").as_deref())
    |         --------------------------------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `str`, found struct `std::string::String`
    |         |
    |         arguments to this function are incorrect
    |
    = note: expected enum `Option<&str>`
               found enum `Option<&std::string::String>`
note: function defined here
   --> /home/epage/src/personal/bat/src/config.rs:92:8
    |
92  | pub fn get_pager_executable(config_pager: Option<&str>) -> Option<String> {
    |        ^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
```

Reply:
Thank you for the detailed explanation. Not to digress too much, but now I'm interested :smile:. What about the additional allocations? I imagine it was previously possible to directly access the (Os)Str-slices inside the environment (`std::env::args`) which are already in memory when the program is started. If every argument is now converted to a `Os(Str)`**`ing`**, wouldn't that lead to a performance impact for programs with *many* arguments (e.g. when doing something like `find … | xargs <myprog>`)? Or is the time for those N allocations insignificant compared to everything else (CLI parsing etc.)

Reply:
> What about the additional allocations? I imagine it was previously possible to directly access the (Os)Str-slices inside the environment (std::env::args) which are already in memory when the program is started. 

`std::env::args_os` returns newly allocated `OsString`s.  clap does take ownership of those but, at the moment, we do not move those directly into `ArgMatches` but will do a clone anyways.  Things get complicate with ownership when you support moving items out of the input, adding items back in, dealing with short flags, and the extra ownership complications caused by needing to use `os_str_bytes`.

I did leave the `ValueParser` trait open to us moving `OsString` into `ArgMatches` if we can do that at a later time.  We can then avoid an extra allocation for `OsString`, `String`, and `PathBuf`.

On a related note, there is an extra source of allocations if you use `ArgGroup`.  An argument inside of an `ArgGroup` will have all of its values cloned into the `ArgGroup`s spot in `ArgMatches`.  I don't think I've even seen this documented anywhere.  I'm changing the behavior to instead put the argument's ID in the `ArgGroup`s match.  This makes it independent of how many values `xargs` passed in, still allows you to look up the value, and also let's you see which group was selected.

Reply:
iirc gumpdrop does not do any extra allocations than what `std::env::args_os` forces on it and [it still has similar performance to clap in the xargs case](https://github.com/rosetta-rs/argparse-rosetta-rs)

Reply:
> std::env::args_os returns newly allocated OsStrings

Oh, my bad. For some reason, I thought that Rust provided direct read-only "`OsStr`" access to the underlying "`const char* argv[]`", like in C/C++, but that's not the case. For some more context on this, see https://github.com/rust-lang/rfcs/issues/862 or https://github.com/rust-lang/rust/issues/47164.

>  Things get complicate with ownership when you support moving items out of the input, adding items back in, dealing with short flags, and the extra ownership complications caused by needing to use os_str_bytes.
> I did leave the ValueParser trait open to us moving OsString into ArgMatches if we can do that at a later time. We can then avoid an extra allocation for OsString, String, and PathBuf.

I see, thank you for the detailed explanation!